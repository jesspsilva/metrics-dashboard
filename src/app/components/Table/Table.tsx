import styled from "styled-components";

import Badge from "@components/Badge/Badge";
import { formatNumber } from "@/app/utils/format-number";

const TableWrapper = styled.table`
  width: 100%;
  border-collapse: collapse;
`;

const TableHead = styled.thead`
  border-bottom: 1px solid var(--light-gray);

  @media (max-width: 768px) {Â´
    border: none;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
  }
`;

const TableHeader = styled.th`
  padding: 20px;

  text-align: left;
  color: var(--gray);
  font-weight: 500;
  font-size: 1rem;
  line-height: 1.5;
`;

const TableRow = styled.tr`
  @media (max-width: 768px) {
    display: block;
    border: 1px solid var(--light-gray);
    box-shadow: 0 2px 6px 0 rgba(0, 0, 0, 0.05);

    &:not(:last-of-type) {
      margin-bottom: 4rem;
      background: var(--white);
    }
  }
`;

const TableData = styled.td`
  font-weight: 400;
  padding: 20px;
  text-align: left;
  font-size: 1rem;
  color: var(--dark-blue);
  line-height: 1.5;

  border-bottom: 1px solid var(--light-gray);

  &:first-child {
    text-transform: capitalize;
  }

  @media (max-width: 768px) {
    display: flex;
    justify-content: space-between;
    gap: 10px;
    padding: 12px;

    font-size: 0.9em;
    text-align: right;

    &::before {
      content: attr(data-label);
      float: left;
      font-weight: bold;
    }

    &:last-child {
      border-bottom: 0;
    }
  }
`;

export default function Table({ data }: { data: MetricsData[] }) {
  return (
    <TableWrapper>
      <TableHead>
        <tr>
          <TableHeader>Label</TableHeader>
          <TableHeader>Value</TableHeader>
          <TableHeader>Description</TableHeader>
          <TableHeader>Category</TableHeader>
        </tr>
      </TableHead>
      <tbody>
        {data.map((item) => (
          <TableRow key={item.id}>
            <TableData data-label="Label">
              {item.label ? item.label : "-"}
            </TableData>
            <TableData data-label="Value">
              {item.value ? formatNumber(item.value, item.type) : "-"}
            </TableData>
            <TableData data-label="Description">
              {item.description ? item.description : "-"}
            </TableData>
            <TableData data-label="Category">
              {item.category ? <Badge label={item.category} /> : "-"}
            </TableData>
          </TableRow>
        ))}
      </tbody>
    </TableWrapper>
  );
}
