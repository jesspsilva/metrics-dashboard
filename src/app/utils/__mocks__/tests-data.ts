export const metricsData = [
  {
    id: "oee",
    label: "oee",
    value: 0.68,
    type: "percentage",
    description: "The overall equipment efficiency in %",
    category: "efficiency",
  },
  {
    id: "sl",
    label: "Speed loss",
    value: -10.5,
    type: "number",
    description: "The line speed loss",
    category: "efficiency",
  },
  {
    id: "cln_shift",
    label: "Cleaning in shift",
    value: 2280,
    type: "secs",
    description:
      "Time spent (in seconds) cleaning the machines during the last shift",
    category: "shift",
  },
  {
    id: "shift_duration",
    label: "Shift duration",
    value: 8,
    type: "hours",
    description: "Last shift duration in hours",
    category: "shift",
  },
  {
    id: "no_value",
    label: "",
    value: null,
    type: "",
    description: "",
    category: "",
  },
  {
    id: "unexplained",
    label: "Unexplained downtime",
    value: 180,
    type: "secs",
    description: "The downtime (in seconds) caused by undocumented reasons",
    category: "downtime",
  },
  {
    id: "mech_problems",
    label: "Mechanical problems",
    value: 1210,
    type: "secs",
    description: "The downtime (in seconds) caused by mechanical problems",
    category: "downtime",
  },
] as MetricsData[];
