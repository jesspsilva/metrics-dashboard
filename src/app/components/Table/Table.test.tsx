import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import Table from "./Table";
import { metricsData as data } from "@/app/utils/__mocks__/tests-data";

describe("Table", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<Table data={data} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the correct number of rows", () => {
    render(<Table data={data} />);
    const rows = screen.getAllByRole("row");
    expect(rows).toHaveLength(data.length + 1);
  });

  it("should render the correct number of columns", () => {
    render(<Table data={data} />);
    const rows = screen.getAllByRole("row");
    const columns = rows[0].querySelectorAll("th");
    expect(columns).toHaveLength(4);
  });

  describe("when table label is available", () => {
    it("should render the label", () => {
      const { container } = render(<Table data={data} />);
      const label = container.querySelectorAll('[data-label="Label"]');
      expect(label[0]).toHaveTextContent("oee");
      expect(label[1]).toHaveTextContent("Speed loss");
      expect(label[2]).toHaveTextContent("Cleaning in shift");
      expect(label[3]).toHaveTextContent("Shift duration");
    });
  });

  describe("when table label is not available", () => {
    it("should render a dash", () => {
      const { container } = render(<Table data={data} />);
      const label = container.querySelectorAll('[data-label="Label"]');
      expect(label[4]).toHaveTextContent("-");
    });
  });

  describe("when table value is a percentage", () => {
    it("should format the number correctly", () => {
      const { container } = render(<Table data={data} />);
      const value = container.querySelectorAll('[data-label="Value"]')[0];
      expect(value).toHaveTextContent("68%");
    });
  });

  describe("when table value is a number", () => {
    it("should format the number correctly", () => {
      const { container } = render(<Table data={data} />);
      const value = container.querySelectorAll('[data-label="Value"]')[1];
      expect(value).toHaveTextContent("-10.5");
    });
  });

  describe("when table value is in seconds", () => {
    it("should format the number correctly", () => {
      const { container } = render(<Table data={data} />);
      const value = container.querySelectorAll('[data-label="Value"]')[2];
      expect(value).toHaveTextContent("2280 s");
    });
  });

  describe("when table value is in hours", () => {
    it("should format the number correctly", () => {
      const { container } = render(<Table data={data} />);
      const value = container.querySelectorAll('[data-label="Value"]')[3];
      expect(value).toHaveTextContent("8 h");
    });
  });

  describe("when table value is not available", () => {
    it("should render a dash", () => {
      const { container } = render(<Table data={data} />);
      const value = container.querySelectorAll('[data-label="Value"]')[4];
      expect(value).toHaveTextContent("-");
    });
  });

  describe("when table description is available", () => {
    it("should render the description", () => {
      const { container } = render(<Table data={data} />);
      const description = container.querySelectorAll(
        '[data-label="Description"]',
      );
      expect(description[0]).toHaveTextContent(
        "The overall equipment efficiency in %",
      );
      expect(description[1]).toHaveTextContent("The line speed loss");
      expect(description[2]).toHaveTextContent(
        "Time spent (in seconds) cleaning the machines during the last shift",
      );
      expect(description[3]).toHaveTextContent("Last shift duration in hours");
    });
  });

  describe("when table description is not available", () => {
    it("should render a dash", () => {
      const { container } = render(<Table data={data} />);
      const description = container.querySelectorAll(
        '[data-label="Description"]',
      );
      expect(description[4]).toHaveTextContent("-");
    });
  });

  describe("when table category is available", () => {
    it("should render the category", () => {
      const { container } = render(<Table data={data} />);
      const category = container.querySelectorAll('[data-label="Category"]');
      expect(category[0].querySelector("span")).toHaveTextContent("efficiency");
      expect(category[1].querySelector("span")).toHaveTextContent("efficiency");
      expect(category[2].querySelector("span")).toHaveTextContent("shift");
      expect(category[3].querySelector("span")).toHaveTextContent("shift");
    });
  });

  describe("when table category is not available", () => {
    it("should render a dash", () => {
      const { container } = render(<Table data={data} />);
      const category = container.querySelectorAll('[data-label="Category"]');
      expect(category[4]).toHaveTextContent("-");
    });
  });
});
