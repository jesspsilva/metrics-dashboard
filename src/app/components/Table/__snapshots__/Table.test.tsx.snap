// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correctly 1`] = `
<DocumentFragment>
  <table
    class="sc-de9c0f69-0 fQQWSd"
  >
    <thead
      class="sc-de9c0f69-1 oYdKO"
    >
      <tr>
        <th
          class="sc-de9c0f69-2 hbnReg"
        >
          Label
        </th>
        <th
          class="sc-de9c0f69-2 hbnReg"
        >
          Value
        </th>
        <th
          class="sc-de9c0f69-2 hbnReg"
        >
          Description
        </th>
        <th
          class="sc-de9c0f69-2 hbnReg"
        >
          Category
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          oee
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          68%
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          The overall equipment efficiency in %
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          <span
            class="sc-fea6eeed-0 jmkfQA bg-emerald-200 border-emerald-800 text-emerald-800"
            label="efficiency"
          >
            efficiency
          </span>
        </td>
      </tr>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          Speed loss
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          -10.5
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          The line speed loss
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          <span
            class="sc-fea6eeed-0 jmkfQA bg-emerald-200 border-emerald-800 text-emerald-800"
            label="efficiency"
          >
            efficiency
          </span>
        </td>
      </tr>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          Cleaning in shift
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          2280 s
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          Time spent (in seconds) cleaning the machines during the last shift
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          <span
            class="sc-fea6eeed-0 jmkfQA bg-sky-200 border-sky-800 text-sky-800"
            label="shift"
          >
            shift
          </span>
        </td>
      </tr>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          Shift duration
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          8 h
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          Last shift duration in hours
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          <span
            class="sc-fea6eeed-0 jmkfQA bg-sky-200 border-sky-800 text-sky-800"
            label="shift"
          >
            shift
          </span>
        </td>
      </tr>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          -
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          -
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          -
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          -
        </td>
      </tr>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          Unexplained downtime
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          180 s
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          The downtime (in seconds) caused by undocumented reasons
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          <span
            class="sc-fea6eeed-0 jmkfQA bg-red-200 border-red-800 text-red-800"
            label="downtime"
          >
            downtime
          </span>
        </td>
      </tr>
      <tr
        class="sc-de9c0f69-3 dLMrcw"
      >
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Label"
        >
          Mechanical problems
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Value"
        >
          1210 s
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Description"
        >
          The downtime (in seconds) caused by mechanical problems
        </td>
        <td
          class="sc-de9c0f69-4 kqUsCt"
          data-label="Category"
        >
          <span
            class="sc-fea6eeed-0 jmkfQA bg-red-200 border-red-800 text-red-800"
            label="downtime"
          >
            downtime
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
